File: LinkList/剑指 Offer 06. 从尾到头打印链表.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */

/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* reversePrint(struct ListNode* head, int* returnSize) {
  struct ListNode* cur = (struct ListNode*)malloc(sizeof(struct ListNode));
  cur = head;
  int len = 0;
  while (cur) {
    len++;
    cur = cur->next;
  }
  int* ans = (int*)malloc(sizeof(int) * len);
  int i = len - 1;
  while (head) {
    ans[i--] = head->val;
    head = head->next;
  }
  *returnSize = len;
  return ans;
}

File: LinkedList链表/剑指 Offer 06. 从尾到头打印链表.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
//  倒序，栈，递归
class Solution {
    public int[] reversePrint(ListNode head) {
        // 如果传入的是ListNode类型，最后就是pop().val
        Deque<Integer> st = new LinkedList<Integer>();
        while(head != null){
            st.push(head.val);
            head = head.next;
        }
        int[] ans = new int[st.size()];
        int l = 0;
        while(!st.isEmpty()){
            ans[l++] = st.pop();
        }
        return ans;
    }
}
